src/datasources/asset/components/AssetQueryEditor.tsx

```diff
@@ -0,0 +1,156 @@
+import React, { useState } from 'react';
+import { QueryEditorProps, SelectableValue, toOption } from '@grafana/data';
+import { InlineField } from 'core/components/InlineField';
+import { AssetDataSource } from "../AssetDataSource";
+import {
+  AssetQuery,
+  EntityType,
+} from '../types';
+import { FloatingError, parseErrorMessage } from "../../../core/errors";
+import { MultiSelect, Select } from "@grafana/ui";
+import { isValidId } from "../../data-frame/utils";
+import { useWorkspaceOptions } from "../../../core/utils";
+import { SystemMetadata } from "../../system/types";
+import _ from "lodash";
+import { useAsync } from "react-use";
+import { selectors } from "../../../test/selectors";
+
+type Props = QueryEditorProps<AssetDataSource, AssetQuery>;
+
+export function AssetQueryEditor({ query, onChange, onRunQuery, datasource }: Props) {
+
+  query = datasource.prepareQuery(query);
+  const workspaces = useWorkspaceOptions(datasource);
+  const [errorMsg, setErrorMsg] = useState<string | undefined>('');
+  const handleError = (error: Error) => setErrorMsg(parseErrorMessage(error));
+
+  const minionIds = useAsync(() => {
+    let filterString = '';
+    if (query.workspace) {
+      filterString += `workspace = "${query.workspace}"`;
+    }
+    return datasource.querySystems(filterString).catch(handleError);
+  }, [query.workspace]);
+
+  const handleQueryChange = (value: AssetQuery, runQuery: boolean): void => {
+    onChange(value);
+    if (runQuery) {
+      onRunQuery();
+    }
+  };
+
+  const onWorkspaceChange = (item?: SelectableValue<string>): void => {
+    if (item?.value) {
+      // if workspace changed, reset Systems and Assets fields
+      handleQueryChange(
+        { ...query, workspace: item.value, minionIds: [] },
+        // do not run query if workspace not changed
+        query.workspace !== item.value
```

Comment:
I think it would make more sense to push this up to the if statement condition, because we don't need to call `onChange` if the workspace is the same.

`if (item?.value && item.value !== query.workspace)`